qlib/TableMapper.qm

```diff
@@ -756,34 +773,62 @@ on_error table_mapper.rollback();
 
             @param rec a hash representing a single input record
 
-            @return if batch data was inserted then a hash (columns) of lists (row data) of all data inserted and potentially returned (in case of sequences) from the database server is returned; if constant mappings are used with batch data, then they are returned as single values assigned to the hash keys
+            @return if batch data was inserted (flushed) then a hash (columns) of lists (row data) of all data inserted and potentially returned (in case of sequences) from the database server is returned; if constant mappings are used with batch data, then they are returned as single values assigned to the hash keys; if nothing was flushed to the database in the call, Qore::NOTHING is returned
+
+            @note this function does not process hashes of lists (bulk format), it expects always single input record on the input
 
-            @throw MAPPER-BATCH-ERROR this exception is thrown if this method is called when the \c "unstable_input" option was given in the constructor
             @throw MISSING-INPUT a field marked mandatory is missing
             @throw STRING-TOO-LONG a field value exceeds the maximum value and the 'trunc' key is not set
             @throw INVALID-NUMBER the field is marked as numeric but the input value contains non-numeric data
+            @throw MAP-ERROR when the function is called wwith hash of lists (bulk) input
         */
         private *hash queueDataIntern(hash rec) {
             if (insert_block == 1) {
                 return insertRow(rec);
             }
 
+            *int rec_list_size = getRecListSize(rec);
+            if (exists rec_list_size) {
+                throw "MAP-ERROR", sprintf("TableMapper::queueDataIntern() does not "
+                        "accept hash of lists - this is likely some internal "
+                        "error. Got %y", rec);
+            }
+
             hash h = mapDataIntern(rec);
 
             # map record data to get the keys for the buffer
+            # FIXME ANOTHER problem: if we already get some thing in hbuf
```

Comment:
this is not a problem; this is correct code.  If there is something already in hbuf, that means there is data queued to be processed and then we add new data to the cached data.  That is now the InboundTableMapper class works so that it has maximum performance by leveraging bulk DML